File: server/engine/actuators/footballActuator.py
import pybullet as p

from server.utilities.distance3D import distance3D
from server.utilities.positionSwap import positionSwap
from server.utilities.rotationCal import getForwardVectorFromYaw, getYaw


# This helper function will find entities by their tag
def find_ent(entities, tag):
    first_matched = None
    for ent in entities.values():
        if ent.tag == tag:
            first_matched = ent.id
            break
    if first_matched:
        return entities[first_matched]
    else:
        return None


def footballActuator(action, agentData, entities, eM, client, indexOfAction):
    agent = entities[agentData.id]
    agent_bullet_id = eM[agentData.id]
    new_state_space_agent = dict(agent.state_space)
    if "Footballer" not in agent.capabilities:
        return

    if "TemporalMemory" in agent.capabilities:
        new_state_space_agent["last_action_index"] = indexOfAction
        if agent.last_action == action:
            new_state_space_agent["last_action_counter"] += 1
        else:
            new_state_space_agent["last_action_counter"] = 1

    if action != "kick":
        agent.last_action = action
        agent.state_space = new_state_space_agent
        return

    ball = find_ent(entities, "ball")
    if ball is None:
        agent.last_action = action
        agent.state_space = new_state_space_agent
        return

    ball_bullet_id = eM[ball.id]
    new_state_space_ball = dict(ball.state)

    agentPos, agentQuat = p.getBasePositionAndOrientation(
        agent_bullet_id, physicsClientId=client
    )
    rot = p.getEulerFromQuaternion(agentQuat, physicsClientId=client)
    yaw = getYaw(rot)
    Rx, Ry = getForwardVectorFromYaw(yaw)

    ballPos, _ = p.getBasePositionAndOrientation(ball_bullet_id, physicsClientId=client)
    dist = distance3D(agentPos, ballPos)
    kickRadius = 2.0
    kickStrength = 15.0

    if kickRadius < dist:
        new_state_space_agent["lastKickSuccess"] = False
        agent.last_action = action
        agent.state_space = new_state_space_agent
        return

    linear_vel, angular_vel = p.getBaseVelocity(ball_bullet_id)
    _, _, lz = linear_vel

    new_linear_vel = [Rx * kickStrength, Ry * kickStrength, lz]

    p.resetBaseVelocity(
        ball_bullet_id,
        linearVelocity=new_linear_vel,
        angularVelocity=angular_vel,
        physicsClientId=client,
    )

    new_state_space_agent["lastKickSuccess"] = True
    new_state_space_agent["previous_distance_ball"] = 1.0

    agent.last_action = action
    agent.state_space = new_state_space_agent

    new_state_space_ball["lastTouchedBy"] = agent.id
    new_state_space_ball["lastTouchedTeam"] = agent.teamId

    ball.state = new_state_space_ball

File: src/engine/runtime/actuators/FootballActuator.js
import { useSceneStore } from "../../../stores/useSceneStore";
import distance3D from "../../utility/3dDistance";
import { getForwardVectorFromYaw } from "../../utility/rotationCal";
import { getIndexOfObs } from "../../utility/getIndex";

function getYawFromRapierBody(body) {
  const q = body.rotation();
  return 2 * Math.atan2(q.y, q.w);
}

export default function footballAdapter(action, agentData, actionSpace) {
  const { updateEntity, entities, updateEntityStat, bodies } =
    useSceneStore.getState();

  const agentBody = bodies[agentData.id];
  const agent = entities[agentData.id];

  const indexOfAction = getIndexOfObs(actionSpace, action);
  const capabilities = agent.capabilities;
  let newStateSpace = { ...agent.state_space };

  if (capabilities.includes("TemporalMemory")) {
    newStateSpace.last_action_index = indexOfAction;
    if (agent.last_action === action) {
      newStateSpace.last_action_counter += 1;
    } else {
      newStateSpace.last_action_counter = 1;
    }
  }

  if (action !== "kick") {
    updateEntity(agent.id, {
      last_action: action,
      state_space: newStateSpace,
    });
    updateEntityStat(agent.id, {
      last_action: action,
      state_space: newStateSpace,
    });
    return;
  }

  if (!agentBody || !agent) return;

  const ballId = Object.keys(entities).find(
    (id) => entities[id].tag === "ball",
  );

  const ballBody = bodies[ballId];

  if (!ballBody) return;

  const agentT = agentBody.translation();
  const ballT = ballBody.translation();

  const agentPos = [agentT.x, agentT.y, agentT.z];
  const ballPos = [ballT.x, ballT.y, ballT.z];

  const dist = distance3D(agentPos, ballPos);
  const kickRadius = agent.settings?.kickRadius ?? 1.2;
  const kickStrength = agent.settings?.kickStrength ?? 15.0;

  if (dist > kickRadius) {
    newStateSpace.lastKickSuccess = false;
    updateEntity(agent.id, {
      last_action: action,
      state_space: newStateSpace,
    });
    updateEntityStat(agent.id, {
      last_action: action,
      state_space: newStateSpace,
    });
    return;
  }

  const yaw = getYawFromRapierBody(agentBody);
  const { x: Rx, z: Rz } = getForwardVectorFromYaw(yaw);

  const currentVel = ballBody.linvel();

  ballBody.setLinvel(
    {
      x: Rx * kickStrength,
      y: currentVel.y,
      z: Rz * kickStrength,
    },
    true,
  );

  newStateSpace.lastKickSuccess = true;

  updateEntity(agent.id, {
    last_action: action,
    state_space: newStateSpace,
  });

  updateEntityStat(agent.id, {
    last_action: action,
    state_space: newStateSpace,
  });

  updateEntity(ballId, {
    state: {
      ...(entities[ballId].state ?? {}),
      lastTouchedBy: agent.id,
      lastTouchedTeam: agent.teamId,
    },
  });
}
